pingcap/ticdc
# kv/client: fix token counter in region scan limiter

### What problem does this PR solve? 

- Fix token is not correctly updated in some corner case
- Add more metrics

### What is changed and how it works?

- Fix token is not correctly updated in some corner case
- Add more metrics
- Fix a bug that context is not used correctly when calling `onRegionFail`, which may lead to region loss, the error region should be re-dispatched, but ignore in error channel. (This Fix has been verified by manual test in the next part)

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
   - Setup a cdc replication cluster with 3 TiKVs, in the upstream prepare more than 10k regions.
   - Reload one of the TiKV
   - The resolved ts of TiCDC should recovery in 2 minutes.

<img src="https://user-images.githubusercontent.com/1527315/121283529-dfb6f780-c90d-11eb-9b7d-b460a7c09fde.png" width=400px>

<img src="https://user-images.githubusercontent.com/1527315/121283561-ef364080-c90d-11eb-9066-a1fcbc0b77c4.png" width=400px>

### Release note

- No release note
